Rudimentary name mangling support

dwblaikie opened the pull request:
This seems to be enough to avoid naming collisions for functions in any of the current test cases (verified by asserting that the name of the `llvm::Function` matches the name passed to create it - not triggering LLVM's numbering that happens when names collide)

It currently implements mangling for namespace scopes, class scopes, and impls.
Nothing generic is mangled yet - haven't looked at how that works, though evidently it's not covered by existing testing, I guess.

Follow-up change will document the current mangling algorithm in `toolchain/docs/lower.md`

dwblaikie committed:
Refactor name mangling into a separate class

In preparation for adding broader name mangling support.

dwblaikie committed:
Add a name mangling prefix.

_C was chosen without special care or consideration - that C++ didn't
choose _C for the itanium mangling probably says it's not the best
letter to pick, but we don't have any particular detail about how
different languages have chosen their mangling prefixes or other
conflict avoidance.

dwblaikie committed:
Namespace mangling

dwblaikie committed:
Mangle class scopes

dwblaikie committed:
Manually update assembly test for name mangling prefix

dwblaikie committed:
Add impl/interface mangling

dwblaikie committed:
Append name components directly

dwblaikie committed:
Mangle impls

* mangle the fully qualified name of the class doing the implementing
* do not mangle the qualifiers of the interfaces yet

dwblaikie committed:
Match parameter names

jonmeow commented on toolchain/lower/mangler.h:
```diff
@@ -0,0 +1,28 @@
+                                       SemIR::NameScopeId name_scope_id)
+      -> void;
+  FileContext& file_context_;
+};
```
```suggestion

class Mangler {
 public:
  explicit Mangler(FileContext& file_context) : file_context_(file_context) {}
  
  auto Mangle(SemIR::FunctionId function_id) -> std::string;

 private:
  auto MangleInverseQualifiedNameScope(bool first_name_component,
                                       llvm::raw_ostream& os,
                                       SemIR::NameScopeId name_scope_id)
            -> void;
            
  FileContext& file_context_;
};

```

Vertical whitespace for:

- Separating the class from the wrapping namespace
- Generally spacing out constructors, functions, and member variables.
    - Generally I'd suggest vertical whitespace except between things like a chain of accessors or variables that don't need comments. (with comments, it can be nice to have a blank line before each comment starts)

jonmeow commented on toolchain/lower/mangler.h:
```diff
@@ -0,0 +1,28 @@
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::Lower {
+class Mangler {
```
Can you add comments to these?

dwblaikie replied:
Sure

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+  }
+  std::string result;
+  llvm::raw_string_ostream os(result);
+  os << "_C";
```
Sorry, I was kind of hoping @chandlerc would review and handle this in particular. I know we'd discussed `_C` as an option, but actually, considering [swift mangling](https://github.com/swiftlang/swift/blob/main/docs/ABI/Mangling.rst) uses `$s`, should we consider a similar `$c`? I think one of the concerns with `_C` might've been too high a risk of conflicting with a C-based name, but maybe using a special character in mangling is a good way to avoid that problem entirely? Or maybe `c.` if we think we can use `.` in names?

dwblaikie replied:
I'm open to other ideas - The "_<upper>" form seems to be dominant, though (LLVM's demangler supports _Z for C++, _R for Rust, and _D for the D programming language).

Just looking at LLVM's demangler makes the fair point that a `.` prefix would be problematic (not that that's what you were suggesting) - because of the leading `.` in ELF symbol names, so the demangler strips those off by default.

Work backwards from _Z and use _Y (or perhaps _X, because X is cool)?

Though I'm really OK with whatever everyone else wants - and wouldn't totally object to using something following swift's pattern. So if $c is best for you, happy to switch to that.

jonmeow replied:
We can separate this out if you want to start with _C as probably good.

I was actually thinking further, and if `C.` or `C$` were an option (upper or lower case), particularly using a non-identifier character as separators instead of prefix, it might both (a) make the name easy to read and (b) offer a name that couldn't conflict with a C name.

But I'm realizing the difference between `_C` and `$c` isn't actually that big for us: we should be suffixing all mangled names with the package scope (e.g., `_CFoo.Main`), which should already avoid name conflicts with C names if we're using `.` separators. So if you don't think `C<char>` is a good choice, let's stick with `_C` for cross-language consistency.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+auto Mangler::Mangle(SemIR::FunctionId function_id) -> std::string {
+  auto& sem_ir = file_context_.sem_ir();
+  auto types = sem_ir.types();
+  auto names = sem_ir.names();
```
These feel a little odd to me as local variables, I don't think we generally do that. (`types` looks unused here) Maybe you could inline them, but since you have the class maybe just add a member helper like `sem_ir() { return file_context_.sem_ir(); }`? That'd mirror what we do in a lot of spots, and would end up looking a lot like what you're doing here.

dwblaikie replied:
Sure - added some helpers.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
```
Please check clang-tidy:
https://github.com/carbon-language/carbon-lang/actions/runs/10624260039/job/29485428614

Note, clang-tidy should show up in vscode if clangd is set up. But you can also run it manually (see the .bazelrc)

dwblaikie replied:
Think I got those addressed.

I am not using vscode at the moment, but really should give it a go - and running it with bazel is a bit tricky since the project isn't clalng-tidy clean at the moment.

jonmeow replied:
I believe the project is clang-tidy clean under LLVM 16 -- we are running it on every commit at that version. If you're having issues, fixes are welcome, but also using LLVM 16 should be clean.

dwblaikie replied:
ah, yep - didn't realize the bazel build used my host tools like that - I'd assumed it was more hermetic. got my local situation version matched so that'll be easier (until I forget to update my PATHs after a reboot between LLVM and Carbon development... )

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+              true, os,
+              sem_ir.classes().Get(opt_class_self->class_id).scope_id);
+        } else {
+          auto builtin_self = types.GetAs<SemIR::BuiltinInst>(impl.self_id);
```
Had you considered modelling this as another `CARBON_KIND_SWITCH` or `Get` + `TryAs` + `As` instead of `TryGetAs` + `GetAs`? In particular, the type store has some indirection that is probably repeated in the `Get`.

dwblaikie replied:
Ah, sure, will do.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+        return;
+      }
+      case CARBON_KIND(SemIR::ClassDecl class_decl): {
+        auto& class_node = sem_ir.classes().Get(class_decl.class_id);
+        os << names.GetFormatted(class_node.name_id);
+        break;
+      }
+      case SemIR::Namespace::Kind: {
+        auto name = names.GetAsStringIfIdentifier(parent.name_id);
+        CARBON_CHECK(name) << "Unexpected special name for namespace: "
+                           << parent.name_id;
```
After our whole discussion, I was thinking you'd do `GetAsStringIfIdentifier` everywhere, were you running into an issue or did I misunderstand?

dwblaikie replied:
Yeah, think I just got a bit confused after stumbling over the NameScope::name_id is only valid for namespaces, and other name_ids you have to go looking in the underlying entity - and then I was looking for how to render those entities' names and came across other APIs and copied those, unsure if/how they were load bearing.

But thanks for the prompt - zooming out again, with the code tidied up a bit, I've been able to use that consistently everywhere, I think. (though not able to share many call sites/deduplicate much code, since the name_ids all come from different places, etc)

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+        return;
+      }
+      case CARBON_KIND(SemIR::ClassDecl class_decl): {
+        auto& class_node = sem_ir.classes().Get(class_decl.class_id);
+        os << names.GetFormatted(class_node.name_id);
```
I think we typically append "_info" for `class` disambiguation (similarly `ns` for namespace). `_node` has a different meaning, since we do have nodes in parse.

```suggestion
        auto& class_info = sem_ir.classes().Get(class_decl.class_id);
        os << names.GetFormatted(class_info.name_id);
```

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
```
I think this is similar and avoids a lookup for the package namespace:

```suggestion
  while (name_scope_id.is_valid() && name_scope_id != SemIR::NameScopeId::Package) {
    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
```

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+        const auto& impl = sem_ir.impls().Get(impl_decl.impl_id);
+        if (auto opt_class_self =
+                types.TryGetAs<SemIR::ClassType>(impl.self_id)) {
+          MangleInverseQualifiedNameScope(
```
We're avoiding recursive functions in order to avoid stack depth issues. Can you build an iterative approach for this?

dwblaikie replied:
Sure thing - think I've got that sorted, and basically meant figuring out mangling qualified interface names in too, so that's added/working.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+          os << builtin_self.builtin_inst_kind.label();
+        }
+        os << ':';
+        auto opt_interface_constraint =
```
What's the "opt_" mean here?

dwblaikie replied:
Optional - not sure if there's another convention here. I'd prefer something at the declaration that indicates the optionality (either using an explicit type instead of `auto` or something in the name like `opt_`)

jonmeow replied:
I think `GetAs` doesn't return an optional.

Note, typically we just don't remark on it. But I think it may not make sense here.

dwblaikie replied:
ah, yeah, must've been some lingering thing. Fixed/removed the prefix.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
```
For my own understanding, why inverse instead of forward?

dwblaikie replied:
Added a comment explaining this - I think someone, maybe @zygoloid, suggested that mangling in inverse order may reduce the incidence of names having a common prefix which might slow down uniqueness tests/comparisons (since the comparison would always have to walk the common prefixes before finding something unique).

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,91 @@
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
```
Had you considered "name_scope" instead of "parent"? What's this the parent of?

dwblaikie replied:
Ah, sure - it's the parent of the previous entity (the loop is looping through parents) - but, yeah, might as well name it to match the name_scope_id.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -238,17 +239,8 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id)
 
-  std::string mangled_name;
-  if (SemIR::IsEntryPoint(sem_ir(), function_id)) {
-    // TODO: Add an implicit `return 0` if `Run` doesn't return `i32`.
```
I wouldn't expect the name mangler to add a return. Maybe this TODO should remain here? Perhaps this code should even keep the `IsEntryPoint` call so that doing so is smooth?

dwblaikie replied:
Rephrased the comment and applied it to the return type computation in the previous statement/just above this line.

The fix isn't really in the name mangling, it's in choosing the return type and in the code generation for the function body. So it seems a bit more accurate having the comment apply to the return type computation - though it's still a fair way from where the other part of the fix has to happen in the code generation.

jonmeow reviewed (commented):
Thanks, I'm glad to see the start on mangling! There are some bits that make sense to me, some questions about how mangling should really work.

For documentation, my sense would be that it's a design choice. If you'd like to write a proposal for that (with alternatives, etc), let me know and I'll help walk you through that. Otherwise, maybe just create a Google Doc explaining it for now? Ideally I'd say to update `toolchain/docs/lower.md`, but unfortunately #4242 is still in review.

dwblaikie committed:
Add vertical whitespace to toolchain/lower/mangler.h

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Rename class_node -> class_info

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Avoid unnecessary package namespace lookup

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Keep a comment about needing to remap the return type for Run with void return.

dwblaikie committed:
Add a comment for MangleInverseQualifiedNameScope

dwblaikie committed:
Rename parent to name_scope

dwblaikie committed:
Add helper functions for containers

dwblaikie committed:
Use sem_ir() helper.

dwblaikie committed:
Switch to a CARBON_KIND_SWITCH

dwblaikie committed:
Add comments for Mangler

dwblaikie committed:
Use data iteration instead of a while loop for name printing.

Laying the foundation for removing the recursion from the mangler.

(also fixes the nested switch type, which didn't compile)

dwblaikie committed:
More data iteration stuff

dwblaikie committed:
Remove recursion and add qualification for impl's interface name

dwblaikie committed:
Remove commented out code, add a helper lambda

dwblaikie committed:
Simplify the code a bit

dwblaikie committed:
Minor simplification

dwblaikie committed:
Fix name to match naming convention

dwblaikie reviewed (commented):
OK - sorry for the drip feed of responses earlier. Working on my github review workflow some more.

I think I've addressed all the comments and things are a bit tidier now.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -233,22 +234,14 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id)
 
   // Compute the return type to use for the LLVM function. If the initializing
   // representation doesn't produce a value, set the return type to void.
+  // TODO: Remap return type to i32 if `Run` doesn't return a value.
```
```suggestion
  // TODO: For the `Run` entry point, remap return type to i32 if it doesn't return a value.
```

Note, I'm really just trying to capture that `Run` means the entry point. I think that was previously apparent next to the `IsEntryPoint` before, but I think it's not as obvious now.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+auto Mangler::Mangle(SemIR::FunctionId function_id) -> std::string {
+  const auto& function = sem_ir().functions().Get(function_id);
+  if (SemIR::IsEntryPoint(sem_ir(), function_id)) {
+    // TODO: Add an implicit `return 0` if `Run` doesn't return `i32`.
```
```suggestion
```

jonmeow commented on toolchain/lower/mangler.h:
```diff
@@ -0,0 +1,48 @@
+
+ private:
+  // Mangle this qualified name with inner scope first, working outwards. This
+  // may reduce the incidence of common prefixes in the name mangling. (ie:
```
```suggestion
  // may reduce the incidence of common prefixes in the name mangling. (i.e.:
```

jonmeow commented on toolchain/lower/mangler.h:
```diff
@@ -0,0 +1,48 @@
+
+// A class for producing mangled (deterministically unique, at least partially
+// human readable) names for externally referenceable entities such as extern
+// functions.
```
```suggestion
// human readable) names for externally referenceable entities such as
// functions.
```

"extern" has what I think is a slightly different meaning in Carbon than in C++, would it make sense to omit it here?

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+  llvm::SmallVector<NameEntry> names_to_render;
+  auto add_scope = [&](SemIR::NameScopeId name_scope_id, char prefix) {
+    if (name_scope_id.is_valid() &&
+        name_scope_id != SemIR::NameScopeId::Package) {
```
Sorry, I was thinking about this a little more and I'm realizing the package name needs to be part of mangled names. That is, in tests, `@_CF` should be `@_CF.Main` because it's valid to have `F()` in different packages.

Would you prefer this to be a TODO, or to address it here? `sem_ir().package_id()` is Invalid for `Main` and would be used directly otherwise. At present, NameScope::is_closed_import indicates a package import, at which point the parent wouldn't be traversed (`_COp.Number:Negate.Core` is an example showing the `Core` package, in tests at present because `Core` is imported).

To be sure, the implication I'm describing is that given:

```
// --- foo.carbon
package Foo;
fn F() {}
// --- bar.carbon
package Bar;
import Foo;
fn G() { return Foo.F(); }
```

You will presently produce `@_CF` for foo.carbon, and `@_CF.Foo` for bar.carbon.

dwblaikie replied:
Ah, yeah, meant to leave a FIXME in there somewhere - but thanks for the pointers and examples explaining how the package naming works - so I've implemented that

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+auto Mangler::MangleInverseQualifiedNameScope(llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  struct NameEntry {
```
I think it'd be helpful to have some implementation comments. For example:

```suggestion
  // Maintain a stack of names for delayed rendering of interface impls.
  struct NameEntry {
```

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  struct NameEntry {
+    SemIR::NameScopeId name_id;
```
```suggestion
    SemIR::NameScopeId name_scope_id;
```

We do also have a `NameId` structure -- I think we tend to call these either `scope_id` or `name_scope_id`

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+        auto interface_type =
+            types().GetAs<SemIR::InterfaceType>(impl.constraint_id);
+        auto interface = sem_ir().interfaces().Get(interface_type.interface_id);
+        names_to_render.push_back({interface.scope_id, ':'});
```
Can you use `add_scope` where you currently use `names_to_render.push_back`? Mainly thinking for consistency, one way other the other.

dwblaikie replied:
I had been using direct push_back when the extra conditionals (valid/package) weren't needed, but now that those conditionals aren't needed at all - I removed the helper entirely and left these all as raw push_back, because that seemed still sufficiently legible to not warrant the extra lambda indirection to me, at least.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+    char prefix;
+  };
+  llvm::SmallVector<NameEntry> names_to_render;
+  auto add_scope = [&](SemIR::NameScopeId name_scope_id, char prefix) {
```
This is a stack, right? Maybe "push" instead of "add"?

dwblaikie replied:
Ended up removing this helper since the conditional stuff went away (between invalid cases not actually happening, and the package cases being supported/happening in the main loop instead)

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+  };
+  llvm::SmallVector<NameEntry> names_to_render;
+  auto add_scope = [&](SemIR::NameScopeId name_scope_id, char prefix) {
+    if (name_scope_id.is_valid() &&
```
Is name_scope_id ever invalid here?

dwblaikie replied:
Nope - I might've got confused from some versions of this, or some code that I was "inspired" by.

Removed.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+    -> void {
+  struct NameEntry {
+    SemIR::NameScopeId name_id;
+    char prefix;
```
Can you document the intended semantics of `prefix`? It looks like `:`, `.`, and `\0` are intended values here -- using `\0` in particular seems very subtle since you seem to be relying on implicit 0-init for that.

dwblaikie replied:
Ah, sure - documented the 3 cases (& also removed dependence on implicit 0-init, FWIW) in the comment. Though not sure if it'd be better to document the cases at their usage, and leave the comment here generically describing any prefix and the null-means-no-prefix case.

jonmeow replied:
I'm fine with this, it fits with top-down reading, and people can always cross-reference uses.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+  auto add_scope = [&](SemIR::NameScopeId name_scope_id, char prefix) {
+    if (name_scope_id.is_valid() &&
+        name_scope_id != SemIR::NameScopeId::Package) {
+      names_to_render.push_back({name_scope_id, prefix});
```
Please use designated initializers for structs. We've gone in that direction for [struct style](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md) because it matches what we'd do in Carbon, and avoids some initialization nuances of C++.

Note, I'd also be agnostic if you switch to a `pair` instead of a struct, since we call out an exception for that as tuple-ish.

dwblaikie replied:
Sure thing, done (kept with a named struct and added designated initializers).

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,99 @@
+
+        auto interface_type =
+            types().GetAs<SemIR::InterfaceType>(impl.constraint_id);
+        auto interface = sem_ir().interfaces().Get(interface_type.interface_id);
```
```suggestion
        const auto& interface = sem_ir().interfaces().Get(interface_type.interface_id);
```

I'm guessing this is just an oversight since it's similar to `impl` above, which you do `const auto&` on.

jonmeow reviewed (commented):
Thanks, the approach is looking good. I did realize we need to handle the package differently, but I'm fine splitting that out as a TODO if you'd rather just finish here. I think the necessary changes won't significantly affect the current logic.

BTW, might be worth adding a brief TODO about the generic/impl mangling you note in the PR description, but I'll leave that to you.

And toolchain/docs/lower.md is now merged for edits, but that can also be done separately.

dwblaikie committed:
Use const ref to avoid accidental copy

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Add package name to mangling

dwblaikie committed:
Rename name_id to name_scope_id

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Add comment

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Clarify entry point return type comment

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Document the 'prefix' member

dwblaikie committed:
Remove out-of-place comment

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Spelling fix

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Rephrase comment, avoiding 'extern' terminology

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Fixes for renamed member

dwblaikie committed:
Add FIXME comment for generics

dwblaikie reviewed (commented):
> Thanks, the approach is looking good. I did realize we need to handle the package differently, but I'm fine splitting that out as a TODO if you'd rather just finish here. I think the necessary changes won't significantly affect the current logic.

Ended up adding package support along the way.

> BTW, might be worth adding a brief TODO about the generic/impl mangling you note in the PR description, but I'll leave that to you.

Ended up implementing the impl qualifier mangling in some of the review iterations, so that's done. I'll leave a FIXME for generic support.

> And toolchain/docs/lower.md is now merged for edits, but that can also be done separately.

Mentioned that in the description, to be done in a follow-up.

dwblaikie committed:
Continue mangling, even once reaching the package name

This had been causing mangling to stop for impls, after reaching the
package level of the first part of the name - missing the second part of
the name.

dwblaikie commented:
(with the fix for the mangling ending too early with package support and impls - now `bazel check //...` with a patch to assert an llvm::Function's name matches the name requested (ie: testing that we never request duplicate names and trigger LLVM's name deduplication support) doesn't fail - which is something :) Not sure if it's worth including that assertion in trunk... I guess it probably is?)

zygoloid commented on toolchain/lower/testdata/impl/extend_impl.carbon:
```diff
@@ -31,20 +31,20 @@ fn InstanceAccess(a: A) -> i32 {
 // CHECK:STDOUT: source_filename = "extend_impl.carbon"
 // CHECK:STDOUT:
-// CHECK:STDOUT: define i32 @F() !dbg !4 {
+// CHECK:STDOUT: define i32 @"_CF.A.Main:I.Main.A.Main"() !dbg !4 {
```
This mangling looks a little strange -- specifically repeating the `.A.Main` part twice. It also looks like the trailing portion might be ambiguous. How should we read this?

dwblaikie replied:
Ah, yep - that's a bug.

Fixed.

The issue was that it was mangling in the parent name scopes, including the parent name scope of wherever the impl was defined - which, if I understand correctly, isn't relevant. (an impl is uniquely defined by its interface and constraint, regardless of the scope in which that impl is defined - ie: two impls of the same interface+constraint, written in different namespaces, are invalid/do collide. (well, I guess there's no syntax to define an impl in a namespace, so it only happens to collide between, I guess, two files both containing an impl or a single file containing an inline impl inside a class definition, and one outside, which fails with:
```
basic.carbon:11:1: ERROR: Redefinition of `impl B as A`.
impl B as A {
^~~~~~~~~~~~~
basic.carbon:6:3: Previous definition was here.
  extend impl as A {
  ^~~~~~~~~~~~~~~~~~
```

Fixed this by adding the lexical parent scope after the switch, and having the impl handling `continue` resulting in the parent scope addition being skipped.

Now there shouldn't be any trailing portion after an impl, it should be `<function name>.<self name>:<constraint name>`.

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -12,40 +12,52 @@ namespace Carbon::Lower {
+    // The prefix emitted before this name component. If '\0', no prefix will be
+    // emitted.
+    // * namespace components are separated by '.'
+    // * the two components of an interface are separated by ':'
```
```suggestion
    // - Namespace components are separated by '.'.
    // - The two components of an interface are separated by ':'.
```

Minor capitalization/punctuation. The '-' is just consistency, we've been using '-' (including md files, where prettier reformats '*' to '-' automatically)

jonmeow reviewed (approved):
I think this looks good now. For zygoloid's comment, if you can address it easily here feel free and I can take another look, but I think it'd also make sense to merge this and work in a separate PR since it'll have more narrow impact. (note, when `impl.self_id` is a class, 3 scopes get added to names_to_render, maybe 1 should be cut)

dwblaikie committed:
Comment formatting/sentencification

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Ignore the scope an impl is defined in

only the constraint and interface uniquely identify an impl

dwblaikie commented:
Very much messed this PR state up... trying to figure out how to fix it :/

dwblaikie committed:
Merge branch 'trunk' into mangle_symbols

dwblaikie committed:
Line wrapping

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,118 @@
+  llvm::SmallVector<NameEntry> names_to_render;
+  names_to_render.push_back({.name_scope_id = name_scope_id, .prefix = '.'});
+  while (!names_to_render.empty()) {
+    auto [name_scope_id, prefix] = names_to_render.back();
+    names_to_render.pop_back();
```
```suggestion
    auto [name_scope_id, prefix] = names_to_render.pop_back_val();
```

jonmeow commented on toolchain/lower/mangler.cpp:
```diff
@@ -0,0 +1,118 @@
+    const auto& name_scope = sem_ir().name_scopes().Get(name_scope_id);
+    CARBON_KIND_SWITCH(sem_ir().insts().Get(name_scope.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        names_to_render.clear();
```
Is this clear needed? Looks like you moved the push to the end, so maybe it's not needed anymore?

Note I have in mind something like:

```
class C {
  extend impl as I {
    fn F() {
      class C { extend impl as I { ... } }
    }
  }
}
```

Which seems like a terrible idea, but at the same time, valid code.

```suggestion
```

dwblaikie replied:
Removed the `clear()` call - thanks for the catch.

Though the example code doesn't compile at the moment (maybe local classes aren't supported yet?
```
basic.carbon:7:7: ERROR: Expected expression.
      class C {
      ^~~~~
```

Though once supported, the mangling code will need to be enhanced to handle intermediate function names, not just at the top level.

Without considering local types, I was thinking about nested types and when dependent types get added to interfaces/impls, that might present similar problems depending on exactly how they work. Then we might have a case where an impl is a scope, though I guess it'd still be the top level scope... hmm, but might've bifurcated earlier on - at that point we might want some way to reference existing entities in the mangling, like Itanium...

jonmeow replied:
Yeah, not yet supported.

jonmeow replied:
I couldn't think of a case with existing code either, due to the restrictions on `impl as`

jonmeow reviewed (approved):
Still LGTM, just pointing out a couple small things.

dwblaikie committed:
Update tests

dwblaikie committed:
Remove unnecessary clear() call

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Simplify code with pop_back_val()

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Remove empty line

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>
